[DTensor] force re-compute sharding when normalized_shape differs in fwd layer norm

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115250

**Summary**:
#114174 did not test the case where `elementwise_affine=False` (i.e. `weight` and `bias` are `None`) and this test would fail due to cached sharding propagation. The difference on sharding prop between these cases is, when `weight` and `bias` are None, the forward layer norm op will be recognized as a "static shape op" and `propagate_op_sharding` will be applied rather than `propagate_op_sharding_non_cached`. A fix is to force re-compute sharding when `normalized_shape` changes by setting op schema's `RuntimeSchemaInfo`'s `static_argnum` to include `normalized_shape` (i.e. 1)

**Test**:
pytest test/distributed/_tensor/test_math_ops.py -s -k layer_norm


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @lucasllc

XilunWu committed:
[DTensor] temporary fix on sharding prop for implicit dynamic shape ops

[ghstack-poisoned]

wanchaol reviewed (changes requested):
I think the right way to do this is to Adjust layernorm foward op `RuntimeSchemaInfo` and make the `static_argnum` to be sth like 4 to force include weight/bias

XilunWu committed:
Update on "[DTensor] temporary fix on sharding prop for implicit dynamic shape ops"


**Summary**:
#114174 did not test the case where `elementwise_affine=False` (i.e. `weight` and `bias` are `None`) and this test would fail due to cached sharding propagation. The difference on sharding prop between these cases is, when `weight` and `bias` are None, the forward layer norm op will be recognized as a "static shape op" and `propagate_op_sharding` will be applied rather than `propagate_op_sharding_non_cached`. I call this type of ops "implicit dynamic shape" and add an ad-hoc logic for them as a temporary fix. wanchaol we can discuss offline to see if there can be a good way to handle this.

**Test**:
pytest test/distributed/_tensor/test_math_ops.py -s -k layer_norm


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 kiukchung lucasllc

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm

XilunWu commented:
@pytorchbot merge -f "unrelated CI error: rocm"
